### DISCUSSION TITLE
Profile update API failing if we are giving values other than BMGS(Board, Medium, GradeLevel, Subject)

### DISCUSSION THREAD

**Discussion by swayangjit** [Sunbird-Lern/Community#122] [Issues] [answered]
We are trying to make the SunbirdED code base independent of BMGS, so while our analysis we figured out if we are sending any value  
other than BMGS the profile update API is throwing an error. The request and response is as follows
**Request:**
```
{
    "userId": "225d0748-e4ae-4c7f-844e-37a92b7af78f",
    "framework": {
        "id": [
            "ap_k-12_1"
        ],
        "category1": [
            "cateroy-1"
        ]
    }
}

```

**Response:**

```
{
    "resmsgid": "9d4c204453c90f245d4ad5bf244e6b52",
    "msgid": "9d4c204453c90f245d4ad5bf244e6b52",
    "err": "UOS_USRUPD0051",
    "status": "FAILED",
    "errmsg": "Unsupported field framework.category1."
}
```

As an adopter how will someone configure these categories for the APIs to support?

@AmiableAnil  @reshmi-nair

**Comment by Krishnaj20** [comment]
@AmiableAnil - any update on this issue?

**Reply by Jayaprakash8887** [reply]
@swayangjit Don't think you are allowed to create your own request body. Request field properties from which data will be read has to remain same. However, values for that property should be able to accept any valid framework terms.

**Reply by Jayaprakash8887** [reply]
{
    "userId": "225d0748-e4ae-4c7f-844e-37a92b7af78f",
    "framework": {
        "id": [
            "ap_k-12_1"
        ],
        "category": [
            "category-1"
        ]
    }
}

**Reply by swayangjit** [reply]
@Jayaprakash8887  Let me try it out and update on the same.

**Reply by Ajoymaity** [reply]
@Jayaprakash8887 @swayangjit category1, category2 and category3 are available in framework1, I try to update user profile with category1 but I'm getting error (unsupported framework.category1/ framework not found) . Framework1 has category1, category2, category3 instead of BMGS.
` "framework": {
            "name": "Framework1",
            "identifier": "framework1",
            "categories": [
                {
                    "identifier": "framework1_category1",
                    "code": "category1",
                    "terms": [
                        {
                            "identifier": "framework1_category1_category1term1",
                            "code": "category1term1",
                            "translations": null,
                            "name": "Category1 Term1",
                            "description": "Category1 Term1",
                            "index": 1,
                            "category": "category1",
                            "status": "Live"
                        }
                    ],
                    "translations": null,
                    "name": "Category1",
                    "description": "Category1",
                    "index": 1,
                    "status": "Live"
                },
                {
                    "identifier": "framework1_category2",
                    "code": "category2",
                    "terms": [
                        {
                            "identifier": "framework1_category2_category2term1",
                            "code": "category2term1",
                            "translations": null,
                            "name": "Category2 Term1",
                            "description": "Category2 Term1",
                            "index": 1,
                            "category": "category2",
                            "status": "Live"
                        },
                        {
                            "identifier": "framework1_category2_category3term1",
                            "code": "category3term1",
                            "translations": null,
                            "name": "Category3 Term1",
                            "description": "Category3 Term1",
                            "index": 2,
                            "category": "category2",
                            "status": "Live"
                        },
                        {
                            "identifier": "framework1_category2_category4term1",
                            "code": "category4term1",
                            "translations": null,
                            "name": "Category4 Term1",
                            "description": "Category4 Term1",
                            "index": 3,
                            "category": "category2",
                            "status": "Live"
                        },
                        {
                            "identifier": "framework1_category2_category5term1",
                            "code": "category5term1",
                            "translations": null,
                            "name": "Category5 Term1",
                            "description": "Category5 Term1",
                            "index": 4,
                            "category": "category2",
                            "status": "Live"
                        }
                    ],
                    "translations": null,
                    "name": "Category2",
                    "description": "Category2",
                    "index": 2,
                    "status": "Live"
                },
                {
                    "identifier": "framework1_category3",
                    "code": "category3",
                    "terms": [
                        {
                            "identifier": "framework1_category3_category3term1",
                            "code": "category3term1",
                            "translations": null,
                            "name": "Category3 Term1",
                            "description": "Category3 Term1",
                            "index": 1,
                            "category": "category3",
                            "status": "Live"
                        }
                    ],
                    "translations": null,
                    "name": "Category3",
                    "description": "Category3",
                    "index": 3,
                    "status": "Live"
                },
                {
                    "identifier": "framework1_category4",
                    "code": "category4",
                    "terms": [
                        {
                            "identifier": "framework1_category4_category4term1",
                            "code": "category4term1",
                            "translations": null,
                            "name": "Category4 Term1",
                            "description": "Category4 Term1",
                            "index": 1,
                            "category": "category4",
                            "status": "Live"
                        }
                    ],
                    "translations": null,
                    "name": "Category4",
                    "description": "Category4",
                    "index": 4,
                    "status": "Live"
                },
                {
                    "identifier": "framework1_category5",
                    "code": "category5",
                    "terms": [
                        {
                            "identifier": "framework1_category5_category5term1",
                            "code": "category5term1",
                            "translations": null,
                            "name": "Category5 Term1",
                            "description": "Category5 Term1",
                            "index": 1,
                            "category": "category5",
                            "status": "Live"
                        }
                    ],
                    "translations": null,
                    "name": "Category5",
                    "description": "Category5",
                    "index": 5,
                    "status": "Live"
                }
            ],
            "code": "framework1",
            "type": "K-12",
            "objectType": "Framework"
        }`
**Request:**
`"request": "request": {
        "userId": "155ce3c5-713e-4749-bc1c-95d09c640914",
        "framework": {
            "id": [
                "framework1"
            ],
            "category1": [
                "Category1 Term1"
            ],
            "category2": [
                "Category2 Term1"
            ],
            "category3": [
                "Category3 Term1"
            ]
        }
    }
      `
**Response:**
`{
    "id": "api.user.update",
    "ver": "v3",
    "ts": "2023-05-24 11:20:58:043+0000",
    "params": {
        "resmsgid": "8d9f1be63451afa782c6b65a16c409d8",
        "msgid": "8d9f1be63451afa782c6b65a16c409d8",
        "err": "UOS_USRUPD0051",
        "status": "FAILED",
        "errmsg": "Unsupported field framework.category1."
    },
    "responseCode": "CLIENT_ERROR",
    "result": {}
}`

**Request**
`"request": {
        "userId": "155ce3c5-713e-4749-bc1c-95d09c640914",
        "framework": {
            "id": [
                "framework1"
            ]
        }
    }`
**Response**
`{
    "id": "api.user.update",
    "ver": "v3",
    "ts": "2023-05-24 11:35:01:852+0000",
    "params": {
        "resmsgid": "7d24e73b440142b538649c9265bbf1b3",
        "msgid": "7d24e73b440142b538649c9265bbf1b3",
        "err": "UOS_USRUPD0032",
        "status": "FAILED",
        "errmsg": "No framework found."
    },
    "responseCode": "RESOURCE_NOT_FOUND",
    "result": {}
}`

**Reply by Jayaprakash8887** [reply]
@reshmi-nair Pls confirm if the above request format is the correct format to update framework to user profile

**Reply by Jayaprakash8887** [reply]
@swayangjit Please let us know if the issue is resolved. Else, please verify all configurations.

1. System settings for "userProfileConfig"
     
![image](https://github.com/Sunbird-Lern/Community/assets/19702246/4eefa7a3-2e17-42c3-bf64-cd17fd34a873)

2. If the framework is published and redis is populated with the framework data (for "framework1")
3. Please check if user-org service is able to connect with learning service - config: "sunbird_api_base_url"

**Reply by Ajoymaity** [reply]
@Jayaprakash8887 @swayangjit I'm looking this issue, actually facing some other issue in dev environment. I'll update you once it's tested

**Reply by Jayaprakash8887** [reply]
@swayangjit Please let us know if your query is resolved.

**Reply by Ajoymaity** [reply]
@swayangjit @Jayaprakash8887 now profile is updating with non BMGS framework after updating System Settings of "userProfileConfig"

**Comment by amitpriyadarshi** [comment, marked as answer]
@swayangjit  As per the discussion with @Ajoymaity  and @Jayaprakash8887 sharing the details for introducing new categories into the framework system and use them.

To introduce any new framework category into the system we need to follow below steps:

1. Add the categories into the `master category list` by calling below api. Currently this API is not exposed, you need call it directly into the `learning service`. Also you need to update the naming convention accordingly in the below request payload. One should be very careful while this API.

```
curl --location 'http://localhost:8080/learning-service/framework/v3/category/master/create' \
--header 'Content-Type: application/json' \
--data '{
    "request": {
        "category": {
            "code": "category1",
            "targetIdFieldName": "targetCategory1Ids",
            "channel": "in.ekstep",
            "searchLabelFieldName": "se_category1s",
            "description": "Category1",
            "name": "Category1",
            "orgIdFieldName": "category1Ids",
            "searchIdFieldName": "se_category1Ids"
        }
    }
}'
```

2. Once category is added into master category list, add those categories into system setting against `userProfileConfig` by calling below API.

- Read API for userProfileConfig:

```
curl --location '{{host}}/api/data/v1/system/settings/get/userProfileConfig' \
--header 'Content-Type: json' \
--header 'Authorization: Bearer <api-key>'
```

**NOTE:** take the `value` data from read call, add the newly introduced categories into them and call update api.

- Update API for userProfileConfig:

```
curl --location '{{host}}/api/data/v1/system/settings/set' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <api-key>' \
--header 'x-authenticated-user-token: <user-token>' \
--data '{
    "request": {
        "id": "userProfileConfig",
        "field": "userProfileConfig",
        "value": "{\"fields\":[\"firstName\",\"lastName\",\"profileSummary\",\"avatar\",\"countryCode\",\"dob\",\"email\",\"gender\",\"grade\",\"language\",\"location\",\"phone\",\"subject\",\"userName\",\"webPages\",\"jobProfile\",\"address\",\"education\",\"skills\",\"badgeAssertions\"],\"publicFields\":[\"firstName\",\"lastName\",\"profileSummary\",\"userName\"],\"privateFields\":[\"email\",\"phone\"],\"csv\":{\"supportedColumns\":{\"NAME\":\"firstName\",\"MOBILE PHONE\":\"phone\",\"EMAIL\":\"email\",\"SCHOOL ID\":\"orgId\",\"USER_TYPE\":\"userType\",\"ROLES\":\"roles\",\"USER ID\":\"userId\",\"SCHOOL EXTERNAL ID\":\"orgExternalId\"},\"outputColumns\":{\"userId\":\"USER ID\",\"firstName\":\"NAME\",\"phone\":\"MOBILE PHONE\",\"email\":\"EMAIL\",\"orgId\":\"SCHOOL ID\",\"orgName\":\"SCHOOL NAME\",\"userType\":\"USER_TYPE\",\"orgExternalId\":\"SCHOOL EXTERNAL ID\"},\"outputColumnsOrder\":[\"userId\",\"firstName\",\"phone\",\"email\",\"organisationId\",\"orgName\",\"userType\",\"orgExternalId\"],\"mandatoryColumns\":[\"firstName\",\"userType\",\"roles\"]},\"read\":{\"excludedFields\":[\"avatar\",\"jobProfile\",\"address\",\"education\",\"webPages\",\"skills\"]},\"framework\":{\"fields\":[\"board\",\"gradeLevel\",\"medium\",\"subject\",\"id\",\"category1\",\"category2\",\"category3\",\"category4\"],\"mandatoryFields\":[\"id\"]}}"
    }
}'
```
**NOTE:** Here newly added categories are added at the bottom.


3. Create the framework with newly introduced categories, their terms and associations. Post framework creation publish the framework. Refer framework APIs document [link](http://docs.sunbird.org/latest/apis/framework/).

4. Once framework is created, add the newly created framework in framework list of the respective channel.

cc: @Krishnaj20

**Reply by Jayaprakash8887** [reply]
@swayangjit Please mark this thread as answered if you do not have further queries
